### ISSUE TITLE
Test last heartbeat details shows up on Activity retry

### ISSUE THREAD

**Issue by lorensr** [temporalio/sdk-typescript#538]
A heartbeat-then-immediately fail test that checks heartbeat details on activity retry

**Comment by cretz** [comment]
@lorensr - Here's the Python test:

```python
async def test_activity_heartbeat_details(
    client: temporalio.client.Client, worker: Worker
):
    async def some_activity() -> str:
        info = temporalio.activity.info()
        count = int(next(iter(info.heartbeat_details))) if info.heartbeat_details else 0
        temporalio.activity.logger.debug(
            "Changing count from %s to %s", count, count + 9
        )
        count += 9
        temporalio.activity.heartbeat(count)
        if count < 30:
            raise RuntimeError("Try again!")
        return f"final count: {count}"

    result = await _execute_workflow_with_activity(
        client,
        worker,
        some_activity,
        retry_max_attempts=4,
    )
    assert result.result == "final count: 36"
```

Basically, I just set the details and add 9 each time and have an expected count. I think the test is important to do this way for a couple of reasons: It specifically tests multiple retries and it makes sure that it does the _exact_ number of retries so a heartbeat isn't missed.

This test is, rarely, giving me errors right now in a racy way, so I am curious if it happens to yours too and if Core is the culprit. Obviously the numbers can be changed to whatever, but it's important to test exact number of retries and have each value build on the previous.

**Comment by bergundy** [comment]
Addressed in #523
